fedimint-api/src/db/mod.rs

```diff
@@ -789,6 +833,60 @@ macro_rules! push_db_key_items {
     };
 }
 
+/// MigrationMap is a BTreeMap that maps DatabaseVersions to async functions. These functions are expected
+/// to "migrate" the database from the keyed DatabaseVersion to DatabaseVersion + 1.
+pub type MigrationMap<'a> = BTreeMap<
+    DatabaseVersion,
+    for<'b> fn(
+        &'b mut DatabaseTransaction<'a>,
+    ) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send + 'b>>,
+>;
+
+/// `apply_migrations` iterates from the on disk database version for the module up to `target_dbversion` and executes all
+/// of the migrations that exist in the `MigrationMap`. Each migration in `MigrationMap` updates the database to have the
+/// correct ondisk structures that the code is expecting. The entire migration process is atomic
+/// (i.e migration from 0->1 and 1->2 happen atomically). This function can be called from any module and as
+/// long as the correct migrations are supplied in `MigrationMap`, the module will be able to read
+/// and write from the database successfully.
+pub async fn apply_migrations<'a>(
+    kind: ModuleKind,
+    db: &'a Database,
+    target_dbversion: DatabaseVersion,
+    migration_map: MigrationMap<'a>,
+) -> Result<(), anyhow::Error> {
+    let mut dbtx = db.begin_transaction().await;
+    let ondisk_version = dbtx.get_value(&DatabaseVersionKey).await?;
+    let db_version = if let Some(ondisk_version) = ondisk_version {
+        let mut current_dbversion = ondisk_version;
+
+        if current_dbversion > target_dbversion {
+            return Err(anyhow::anyhow!(format!(
+                "On disk database version for module {kind} was higher than the code database version."
+            )));
+        }
+
+        while current_dbversion < target_dbversion {
+            if let Some(migration) = migration_map.get(&current_dbversion) {
```

Comment:
Shouldn't we require migration for every version bump? After all - if there was a version bump, there had to be a reason for it.

Reply:
That's a good point. I can add a panic which should break our tests if that ever happens.

Reply:
Added `panic!` if there is no migration for a db version